library/unwind/src/wasm.rs

```diff
@@ -0,0 +1,56 @@
+//! A shim for libunwind implemented in terms of the native wasm `throw` instruction.
+
+#![allow(nonstandard_style)]
+
+#[repr(C)]
+#[derive(Debug, Copy, Clone, PartialEq)]
+pub enum _Unwind_Reason_Code {
+    _URC_NO_REASON = 0,
+    _URC_FOREIGN_EXCEPTION_CAUGHT = 1,
+    _URC_FATAL_PHASE2_ERROR = 2,
+    _URC_FATAL_PHASE1_ERROR = 3,
+    _URC_NORMAL_STOP = 4,
+    _URC_END_OF_STACK = 5,
+    _URC_HANDLER_FOUND = 6,
+    _URC_INSTALL_CONTEXT = 7,
+    _URC_CONTINUE_UNWIND = 8,
+    _URC_FAILURE = 9, // used only by ARM EHABI
+}
+pub use _Unwind_Reason_Code::*;
+
+pub type _Unwind_Exception_Class = u64;
+pub type _Unwind_Word = *const u8;
+
+pub const unwinder_private_data_size: usize = 2;
+
+#[repr(C)]
+pub struct _Unwind_Exception {
+    pub exception_class: _Unwind_Exception_Class,
+    pub exception_cleanup: _Unwind_Exception_Cleanup_Fn,
+    pub private: [_Unwind_Word; unwinder_private_data_size],
+}
+
+pub type _Unwind_Exception_Cleanup_Fn =
+    Option<extern "C" fn(unwind_code: _Unwind_Reason_Code, exception: *mut _Unwind_Exception)>;
+
+pub unsafe fn _Unwind_DeleteException(exception: *mut _Unwind_Exception) {
+    if let Some(exception_cleanup) = unsafe { (*exception).exception_cleanup } {
+        exception_cleanup(_URC_FOREIGN_EXCEPTION_CAUGHT, exception);
+    }
+}
+
+pub unsafe fn _Unwind_RaiseException(exception: *mut _Unwind_Exception) -> _Unwind_Reason_Code {
+    extern "C" {
+        /// LLVM lowers this intrinsic to the `throw` instruction.
+        // FIXME(coolreader18): move to stdarch
+        #[link_name = "llvm.wasm.throw"]
+        fn wasm_throw(tag: i32, ptr: *mut u8) -> !;
+    }
+
+    // The wasm `throw` instruction takes a "tag", which differentiates certain
+    // types of exceptions from others. LLVM currently just identifies these
+    // via integers, with 0 corresponding to C++ exceptions and 1 to C setjmp()/longjmp().
+    // Ideally, we'd be able to choose something unique for Rust, but for now,
+    // we pretend to be C++ and implement the Itanium exception-handling ABI.
+    wasm_throw(0, exception.cast())
```

Comment:
This probably needs `#[cfg(panic = "unwind")]` and an `core::intrinsics::abort()` call otherwise. Or does `llvm.wasm.throw` already codegen as abort when the exception-handling feature is disabled?

Reply:
It codegens as throw, I think, but that's fine since this is only pulled in by libpanic_unwind. Unless you mean so that throw never gets codegenned except for with -Zbuild-std? I was trying to figure out how to feature-gate -Cpanic=unwind on wasm, but I don't know if there's a good way to do that, so maybe this would be the better solution. I guess the alternative is just stabilizing it now, but the proposal is still in stage 3, though it is enabled without a flag in all major browsers (FF, Chrome, Safari), so it's not very likely for it to change very much.

Reply:
libstd directly depends on the unwind crate. As such it would pull in this function even when not linking libpanic_unwind. Note that because of the wasm validation everything has to be compiled with exception handling disabled if you want to work on a wasm runtime without exception support. You can't just link against panic_abort instead of panic_unwind. Because of this I think the only option for the time being is to disable exception handling support entirely in the standard library we distribute and require `-Zbuild-std` if you want to use exception handling support.